File: src/backend/app/schemas/agent_task.py
"""Public, project-scoped Agent Task projection contracts.

These models are a read model over canonical lifecycle evidence. They are not
persisted and never become a second workflow state machine.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from src.backend.app.schemas.agent_harness import AgentHarnessSummary

from src.backend.app.schemas.node_contract import CapabilityLevel

AgentTaskPublicState = Literal[
    "preparing",
    "waiting_for_user",
    "running",
    "needs_attention",
    "completed",
]
AgentTaskOutcome = Literal["succeeded", "partial", "failed", "canceled", "indeterminate"]
AgentTaskNextActionType = Literal[
    "none",
    "provide_input",
    "revise_goal",
    "answer_science_decision",
    "approve_execution",
    "approve_recovery",
    "review_results",
    "view_attention",
    "contact_support",
]
AgentTaskProgressPhase = Literal[
    "context",
    "planning",
    "plan_ready",
    "data_preparation",
    "execution",
    "validation",
    "recovery",
    "complete",
]
AgentTaskEvidenceType = Literal[
    "task_details",
    "reviewed_plan",
    "execution_ticket",
    "run",
    "observation",
    "goal_evaluation",
    "artifact",
    "validation",
    "provenance",
    "audit",
    "diagnosis",
    "recovery",
]
AgentTaskEventSource = Literal[
    "lifecycle",
    "reviewed_plan",
    "ticket",
    "run",
    "observation",
    "goal_evaluation",
    "diagnosis",
    "recovery",
    "artifact",
]


class AgentTaskNextAction(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    type: AgentTaskNextActionType
    title: str
    description: str | None = None
    requires_user: bool
    decision_batch_id: str | None = None
    disabled_reason: str | None = None


class AgentTaskProgress(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    phase: AgentTaskProgressPhase
    percent: int | None = Field(default=None, ge=0, le=100)
    completed_subjects: int | None = Field(default=None, ge=0)
    failed_subjects: int | None = Field(default=None, ge=0)
    excluded_subjects: int | None = Field(default=None, ge=0)
    total_subjects: int | None = Field(default=None, ge=0)


class AgentTaskDecisionOption(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    id: str
    label: str
    description: str
    recommended: bool = False


class AgentTaskDecision(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    item_id: str
    kind: Literal[
        "missing_input",
        "goal_revision",
        "subject_id",
        "atlas",
        "global_signal_regression",
        "repetition_time",
        "template",
        "overwrite",
        "experimental_backend",
        "other",
    ]
    question: str
    impact: str
    options: tuple[AgentTaskDecisionOption, ...] = ()
    recommended_option: str | None = None
    source: Literal["planner", "memory_suggestion"] = "planner"
    memory_id: str | None = None
    recommendation_source: str | None = None
    answer_type: Literal["option", "text"] = "option"
    required: bool = True
    evidence_refs: tuple[str, ...] = ()


class AgentTaskDecisionBatch(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    batch_id: str
    evidence_snapshot_hash: str
    plan_hash_before: str | None = None
    expires_at: datetime


class AgentTaskApprovalSection(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    id: str
    title: str
    summary: str
    warnings: tuple[str, ...] = ()


class AgentTaskApprovalSummary(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    summary_hash: str
    goal: str
    dataset_summary: str
    execution_summary: str
    write_roots: tuple[str, ...]
    rawdata_read_only: bool = True
    external_tools: tuple[str, ...] = ()
    limitations: tuple[str, ...] = ()
    science_changes: tuple[str, ...] = ()
    memory_context_hash: str | None = None
    memory_refs: tuple[dict[str, object], ...] = ()
    memory_influence_summary: tuple[str, ...] = ()
    planning_inputs_hash: str | None = None
    revision_no: int | None = Field(default=None, ge=1)
    parent_reviewed_plan_id: str | None = None
    parent_plan_hash: str | None = None
    revision_reason: str | None = None
    sections: tuple[AgentTaskApprovalSection, ...] = ()
    expires_at: datetime | None = None


class AgentTaskArtifactSummary(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    artifact_id: str
    artifact_type: str
    label: str
    uri: str
    checksum: str | None = None
    capability_level: CapabilityLevel
    reload_status: Literal["not_checked", "passed", "failed", "unavailable"]


class AgentTaskResultSummary(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    outcome: AgentTaskOutcome
    title: str
    summary: str
    qc_summary: str | None = None
    completed_subjects: int | None = Field(default=None, ge=0)
    failed_subjects: int | None = Field(default=None, ge=0)
    excluded_subjects: int | None = Field(default=None, ge=0)
    total_subjects: int | None = Field(default=None, ge=0)
    limitations: tuple[str, ...] = ()
    recommended_action: str | None = None
    artifacts: tuple[AgentTaskArtifactSummary, ...] = ()


class AgentResultCriterion(BaseModel):
    """Read-only criterion outcome copied from the deterministic evaluator."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    criterion_id: str
    status: Literal["passed", "failed", "indeterminate"]
    reason_code: str
    evidence_ids: tuple[str, ...] = ()


class AgentResultExplanation(BaseModel):
    """Structured result explanation; generated prose never controls the outcome."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    outcome: AgentTaskOutcome
    completed_subjects: int | None = Field(default=None, ge=0)
    failed_subjects: int | None = Field(default=None, ge=0)
    excluded_subjects: int | None = Field(default=None, ge=0)
    total_subjects: int | None = Field(default=None, ge=0)
    artifact_refs: tuple[AgentTaskArtifactSummary, ...] = ()
    criteria: tuple[AgentResultCriterion, ...] = ()
    limitations: tuple[str, ...] = ()
    recommended_action: str | None = None
    generated_text: str | None = Field(default=None, max_length=2048)
    generated_text_status: Literal["not_requested", "accepted", "conflict_rejected"] = "not_requested"


class AgentTaskRecoverySummary(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    proposal_id: str
    diagnosis: str
    affected_subjects: tuple[str, ...] = ()
    recommended_action: str
    untouched_scope: tuple[str, ...] = ()
    requires_new_plan: bool
    approval_summary_hash: str | None = None


class AgentTaskEvidenceLink(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    id: str
    type: AgentTaskEvidenceType
    label: str
    uri: str
    available: bool


class AgentTaskBackendSelection(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    requested: str
    selected: str | None = None
    fallback_reason: str | None = None


class AgentTaskTechnicalDetails(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    lifecycle_id: str
    internal_state: str
    reviewed_plan_id: str | None = None
    plan_hash: str | None = None
    planning_inputs_hash: str | None = None
    plan_revision_no: int | None = Field(default=None, ge=1)
    parent_reviewed_plan_id: str | None = None
    parent_plan_hash: str | None = None
    revision_reason: str | None = None
    evidence_snapshot_hash: str | None = None
    goal_contract_id: str | None = None
    goal_hash: str | None = None
    ticket_id: str | None = None
    run_id: str | None = None
    observation_id: str | None = None
    evaluation_id: str | None = None
    backend: AgentTaskBackendSelection | None = None
    node_ids: tuple[str, ...] = ()
    memory_context_hash: str | None = None
    memory_refs: tuple[dict[str, object], ...] = ()
    memory_retrieval_policy_version: str | None = None
    memory_status: Literal["disabled", "enabled", "partial"] | None = None
    memory_used_bytes: int | None = Field(default=None, ge=0)
    memory_omitted_count: int | None = Field(default=None, ge=0)
    memory_warnings: tuple[str, ...] = ()
    memory_available: bool | None = None
    memory_generate_enabled: bool | None = None
    memory_use_enabled: bool | None = None


class AgentTaskResponse(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    schema_version: Literal[1] = 1
    task_id: str
    project_id: str
    state: AgentTaskPublicState
    outcome: AgentTaskOutcome | None = None
    goal_summary: str
    current_action: str
    next_action: AgentTaskNextAction
    progress: AgentTaskProgress
    decisions: tuple[AgentTaskDecision, ...] = ()
    decision_batch: AgentTaskDecisionBatch | None = None
    approval_summary: AgentTaskApprovalSummary | None = None
    result_summary: AgentTaskResultSummary | None = None
    result_explanation: AgentResultExplanation | None = None
    recovery: AgentTaskRecoverySummary | None = None
    evidence_links: tuple[AgentTaskEvidenceLink, ...] = ()
    technical_details: AgentTaskTechnicalDetails | None = None
    harness_summary: AgentHarnessSummary | None = None
    created_at: datetime
    updated_at: datetime


class AgentTaskListResponse(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    schema_version: Literal[1] = 1
    items: tuple[AgentTaskResponse, ...]
    total: int = Field(ge=0)


class AgentTaskEvent(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    event_id: str
    task_id: str
    project_id: str
    source: AgentTaskEventSource
    type: str
    occurred_at: datetime
    title: str
    summary: str
    evidence_uri: str | None = None


class AgentTaskEventPage(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    schema_version: Literal[1] = 1
    items: tuple[AgentTaskEvent, ...]
    next_cursor: str | None = None


class CreateAgentTaskRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    goal: str = Field(min_length=1)
    command_id: str = Field(min_length=1)
    actor: str = Field(min_length=1)


class DecisionAnswer(BaseModel):
    model_config = ConfigDict(extra="forbid")

    item_id: str = Field(min_length=1)
    value: str = Field(min_length=1)


class AnswerAgentTaskRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    batch_id: str = Field(min_length=1)
    answers: tuple[DecisionAnswer, ...] = Field(min_length=1, max_length=6)
    command_id: str = Field(min_length=1)
    actor: str = Field(min_length=1)


class ApproveAgentTaskRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    approval_summary_hash: str = Field(min_length=1)
    command_id: str = Field(min_length=1)


class CancelAgentTaskRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    command_id: str = Field(min_length=1)
    actor: str = Field(min_length=1)
    reason: str | None = None


class ApproveAgentTaskRecoveryRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    command_id: str = Field(min_length=1)

File: src/frontend/src/lib/types/agentTask.ts
export type AgentTaskPublicState =
  | "preparing"
  | "waiting_for_user"
  | "running"
  | "needs_attention"
  | "completed";

export type AgentTaskOutcome = "succeeded" | "partial" | "failed" | "canceled" | "indeterminate";

export type AgentTaskNextActionType =
  | "none"
  | "provide_input"
  | "revise_goal"
  | "answer_science_decision"
  | "approve_execution"
  | "approve_recovery"
  | "review_results"
  | "view_attention"
  | "contact_support";

export type AgentTaskProgressPhase =
  | "context"
  | "planning"
  | "plan_ready"
  | "data_preparation"
  | "execution"
  | "validation"
  | "recovery"
  | "complete";

export type AgentTaskDecisionKind =
  | "missing_input"
  | "goal_revision"
  | "subject_id"
  | "atlas"
  | "global_signal_regression"
  | "repetition_time"
  | "template"
  | "overwrite"
  | "experimental_backend"
  | "other";

export type AgentTaskEvidenceType =
  | "task_details"
  | "reviewed_plan"
  | "execution_ticket"
  | "run"
  | "observation"
  | "goal_evaluation"
  | "artifact"
  | "validation"
  | "provenance"
  | "audit"
  | "diagnosis"
  | "recovery";

export type AgentTaskEventSource =
  | "lifecycle"
  | "reviewed_plan"
  | "ticket"
  | "run"
  | "observation"
  | "goal_evaluation"
  | "diagnosis"
  | "recovery"
  | "artifact";

export type AgentTaskBackendSelection = {
  requested: "auto" | "cpu" | "gpu" | string;
  selected: "cpu" | "gpu" | string | null;
  fallback_reason: string | null;
};

export type AgentTaskDecisionOption = {
  id: string;
  label: string;
  description: string;
  recommended: boolean;
};

export type AgentTaskDecision = {
  item_id: string;
  kind: AgentTaskDecisionKind;
  question: string;
  impact: string;
  options: AgentTaskDecisionOption[];
  recommended_option: string | null;
  source?: "planner" | "memory_suggestion";
  memory_id?: string | null;
  recommendation_source?: string | null;
  answer_type: "option" | "text";
  required: boolean;
  evidence_refs: string[];
};

export type AgentTaskDecisionBatch = {
  batch_id: string;
  evidence_snapshot_hash: string;
  plan_hash_before: string | null;
  expires_at: string;
};

export type AgentTaskNextAction = {
  type: AgentTaskNextActionType;
  title: string;
  description: string | null;
  requires_user: boolean;
  decision_batch_id: string | null;
  disabled_reason: string | null;
};

export type AgentTaskProgress = {
  phase: AgentTaskProgressPhase;
  percent: number | null;
  completed_subjects: number | null;
  failed_subjects: number | null;
  excluded_subjects: number | null;
  total_subjects: number | null;
};

export type AgentTaskApprovalSection = {
  id: string;
  title: string;
  summary: string;
  warnings: string[];
};

export type AgentTaskApprovalSummary = {
  summary_hash: string;
  goal: string;
  dataset_summary: string;
  execution_summary: string;
  write_roots: string[];
  rawdata_read_only: boolean;
  external_tools: string[];
  limitations: string[];
  science_changes: string[];
  memory_context_hash?: string | null;
  memory_refs?: Record<string, unknown>[];
  memory_influence_summary?: string[];
  planning_inputs_hash?: string | null;
  revision_no?: number | null;
  parent_reviewed_plan_id?: string | null;
  parent_plan_hash?: string | null;
  revision_reason?: string | null;
  sections: AgentTaskApprovalSection[];
  expires_at: string | null;
};

export type AgentTaskArtifactSummary = {
  artifact_id: string;
  artifact_type: string;
  label: string;
  uri: string;
  checksum: string | null;
  capability_level: "unavailable" | "scaffolded" | "metadata_only" | "computed" | "validated";
  reload_status: "not_checked" | "passed" | "failed" | "unavailable";
};

export type AgentTaskResultSummary = {
  outcome: AgentTaskOutcome;
  title: string;
  summary: string;
  qc_summary: string | null;
  completed_subjects: number | null;
  failed_subjects: number | null;
  excluded_subjects: number | null;
  total_subjects: number | null;
  limitations: string[];
  recommended_action: string | null;
  artifacts: AgentTaskArtifactSummary[];
};

export type AgentResultCriterion = {
  criterion_id: string;
  status: "passed" | "failed" | "indeterminate";
  reason_code: string;
  evidence_ids: string[];
};

export type AgentResultExplanation = {
  outcome: AgentTaskOutcome;
  completed_subjects: number | null;
  failed_subjects: number | null;
  excluded_subjects: number | null;
  total_subjects: number | null;
  artifact_refs: AgentTaskArtifactSummary[];
  criteria: AgentResultCriterion[];
  limitations: string[];
  recommended_action: string | null;
  generated_text: string | null;
  generated_text_status: "not_requested" | "accepted" | "conflict_rejected";
};

export type AgentTaskRecoverySummary = {
  proposal_id: string;
  diagnosis: string;
  affected_subjects: string[];
  recommended_action: string;
  untouched_scope: string[];
  requires_new_plan: boolean;
  approval_summary_hash: string | null;
};

export type AgentTaskEvidenceLink = {
  id: string;
  type: AgentTaskEvidenceType;
  label: string;
  uri: string;
  available: boolean;
};

export type AgentTaskTechnicalDetails = {
  lifecycle_id: string;
  internal_state: string;
  reviewed_plan_id: string | null;
  plan_hash: string | null;
  planning_inputs_hash?: string | null;
  plan_revision_no?: number | null;
  parent_reviewed_plan_id?: string | null;
  parent_plan_hash?: string | null;
  revision_reason?: string | null;
  evidence_snapshot_hash?: string | null;
  goal_contract_id: string | null;
  goal_hash: string | null;
  ticket_id: string | null;
  run_id: string | null;
  observation_id: string | null;
  evaluation_id: string | null;
  backend: AgentTaskBackendSelection | null;
  node_ids: string[];
  memory_context_hash?: string | null;
  memory_refs?: Record<string, unknown>[];
  memory_retrieval_policy_version?: string | null;
  memory_status?: "disabled" | "enabled" | "partial" | null;
  memory_used_bytes?: number | null;
  memory_omitted_count?: number | null;
  memory_warnings?: string[];
  memory_available?: boolean | null;
  memory_generate_enabled?: boolean | null;
  memory_use_enabled?: boolean | null;
};

export type AgentHarnessStatus =
  | "READY"
  | "RUNNING"
  | "WAITING_FOR_USER"
  | "FINISHED"
  | "STOPPED"
  | "FAILED";

export type AgentHarnessSummary = {
  status: AgentHarnessStatus;
  model_calls_used: number;
  model_calls_limit: number;
  action_proposals_used: number;
  action_proposals_limit: number;
  steps_used: number;
  steps_limit: number;
  repairs_used: number;
  repairs_limit: number;
  recovery_attempts_used: number;
  recovery_attempts_limit: number;
  input_tokens_used: number | null;
  input_tokens_limit: number | null;
  output_tokens_used: number | null;
  output_tokens_limit: number | null;
  actual_provider: string | null;
  next_step: string | null;
  terminal_reason: string | null;
  latest_step_id: string | null;
  latest_step_summary: string | null;
  last_wake_reason: string | null;
  yield_count: number;
  fallback_from: string | null;
  fallback_to: string | null;
  fallback_reason: string | null;
};

export type AgentTaskResponse = {
  schema_version: 1;
  task_id: string;
  project_id: string;
  state: AgentTaskPublicState;
  outcome: AgentTaskOutcome | null;
  goal_summary: string;
  current_action: string;
  next_action: AgentTaskNextAction;
  progress: AgentTaskProgress;
  decisions: AgentTaskDecision[];
  decision_batch: AgentTaskDecisionBatch | null;
  approval_summary: AgentTaskApprovalSummary | null;
  result_summary: AgentTaskResultSummary | null;
  result_explanation?: AgentResultExplanation | null;
  recovery: AgentTaskRecoverySummary | null;
  evidence_links: AgentTaskEvidenceLink[];
  technical_details: AgentTaskTechnicalDetails | null;
  harness_summary?: AgentHarnessSummary | null;
  created_at: string;
  updated_at: string;
};

export type AgentTaskListResponse = {
  schema_version: 1;
  items: AgentTaskResponse[];
  total: number;
};

export type AgentTaskEvent = {
  event_id: string;
  task_id: string;
  project_id: string;
  source: AgentTaskEventSource;
  type: string;
  occurred_at: string;
  title: string;
  summary: string;
  evidence_uri: string | null;
};

export type AgentTaskEventPage = {
  schema_version: 1;
  items: AgentTaskEvent[];
  next_cursor: string | null;
};

export type CreateAgentTaskRequest = {
  goal: string;
  command_id: string;
  actor: string;
};

export type AnswerAgentTaskRequest = {
  batch_id: string;
  answers: { item_id: string; value: string }[];
  command_id: string;
  actor: string;
};

export type ApproveAgentTaskRequest = {
  approval_summary_hash: string;
  command_id: string;
};

export type CancelAgentTaskRequest = {
  command_id: string;
  actor: string;
  reason?: string;
};

export type ApproveAgentTaskRecoveryRequest = {
  command_id: string;
};
